### ISSUE TITLE
Segmentation results

### ISSUE THREAD

**Issue by adaa0164** [martinkersner/train-CRF-RNN#9]
Hello,
Thanks for this code, it is very helpful.

I downloaded your code and I followed exactly the procedure that you describe.
I trained until 20000 iterations, and loss was converged like your result.

However, the outputs of segmentor is bad.
I attached outputs to following site.
[label_2007_002094.png](https://www.dropbox.com/s/0lug3w3qhldvmuz/label_2007_002094.png?dl=0)
[label_2007_005844.png](https://www.dropbox.com/s/k3ytkw6foop30ap/label_2007_005844.png?dl=0)
[label_2008_007811.png](https://www.dropbox.com/s/cj8j4u1bxmmp9jk/label_2008_007811.png?dl=0)

What is the cause of these results ?

Thanks!

**Comment by martinkersner** [comment]
Hi @adaa0164!

According to [reply](https://github.com/torrvision/crfasrnn/issues/23) from original CRF-RNN, even 80 thousand iterations might not be sufficient. The referred plot is just illustrating the loss decrease.

I have tried to train for 130+ thousand iterations with three classes and also didn't get much better results. If you use new version of code which contains the [script for testing current model](https://github.com/martinkersner/train-CRF-RNN/blob/master/test_model.py) during training, you will see that when loss seems to be converged, mean IU is only about 45. And it is not enough to provide good results with such model.

Honestly, I don't know where could be a problem, however, I have few tips:
* Something is wrong in training pipeline which I have prepared.
* Not enough data for training only tree classes, objects cover small part of images.
* Don't use SOFTMAX_LOSS layer. From evaluation plots can be seen that loss and the mean IU correlate only loosely. Network doesn't get enough information about error that segmentation makes.
* Missing information about training from [CRF-RNN](https://github.com/torrvision/crfasrnn/).

If you find out what causes the problem of poor segmentation I would be really interested. I am also still working on it.

Martin

**Comment by pierrebaque** [comment]
Hello,

I have the same problem currently, I don't manage to get reasonable results. Did you find out what was the issue?

Best,

Pierre

**Comment by martinkersner** [comment]
Hi!

I haven't found it yet and since I don't know about anybody who has been able to get good results using provided CRF-RNN network, I moved on different segmentation architecture - [DeepLab](https://bitbucket.org/deeplab/deeplab-public). According to paper, DeepLab network achieves comparable results (CRF-RNN is only 0.4% better on PASCAL VOC 2012 test set), moreover training and converging is much faster. They modified size of convolutional filters and computation of convolution. It results in faster training and allows to work with much larger batch size. Their following paper enables to use even weakly annotated ground truth labels.

I tried to [train](https://github.com/martinkersner/train-DeepLab) using PASCAL VOC 2012 dataset and after 6000 thousand iterations without CRF I obtained mean IU 66.7 on val set.

Cheers,

Martin

**Comment by pierrebaque** [comment]
Hey Martin,

Thanks for your answer. Therefore, in the post (https://github.com/torrvision/crfasrnn/issues/28), when you say that you get "better results", it is not compared to what the guys report in the paper, right? It is just bad results which are better than other bad results?

I used DeepLab as well, it works very well. And part of the work I am doing is trying to understand why (if?) end to end training works better.

Thank you

**Comment by martinkersner** [comment]
Exactly, I was just comparing poor results. For both sets of training CRF parameters I didn't get any good segmentation results. 

I guess it should work better, but that's hard to say when we can't make the training working. However,  maybe soon we could get official training scripts for CRF-RNN. @bittnt once mentioned in comment that they are planning to release training scripts soon, but then later on he edited that comment and deleted the mention about training scripts.

**Comment by bittnt** [comment]
Hi, @martinkersner Glad to know that you have made this repository. 

To get better performance on your specific dataset, I think you should first to train the FCN-8 a bit, making sure it is a reasonably good starting point for end-to-end training CRF-RNN. For training CRF-RNN layers, you should plug in the CRF-RNN layers and start to monitor both testing loss and mean IU on every few thousands iterations. You should expect mean IU going down, at least better than the FCN-8 alone. This indicates you are on the right track. We still plan to release the training scripts.

Deeplab does indeed have faster trainable FCN network, and they provide CRF-based postprocessing. Combining that with CRF-RNN, you should expect better performance as well.

**Comment by pierrebaque** [comment]
Hi @bittnt,

Could you maybe provide the pretrained version of FCN-8 that you used as a starting point?
When can we expect the training scripts to be released?

Thanks

Pierre

**Comment by ahundt** [comment]
Probably this
https://gist.github.com/longjon/1bf3aa1e0b8e788d7e1d#file-readme-md

Found here:
https://github.com/BVLC/caffe/wiki/Model-Zoo

**Comment by pierrebaque** [comment]
OK, that is what I used. There must be an issue somewhere in the way I use the CRF part since that the initial results are far from being as good as those of FCN-8.
@martinkersner do you use it as well?

Thank you everyone for your help anyway.

**Comment by martinkersner** [comment]
@pierrebaque Yes, that's what I used.

**Comment by pierrebaque** [comment]
Hello @martinkersner ,

I think I partially solved the issue. In you TVG_CRFRNN_COCO_VOC_TRAIN_20_CLASSES.prototxt and TVG_CRFRNN_COCO_VOC_TEST_20_CLASSES.prototxt, all the names have "-ft" at the end. 
It is not the case in https://gist.github.com/longjon/1bf3aa1e0b8e788d7e1d#file-readme-md, therefore, the initialisation is not properly done.

When I remove the "-ft" the initialisation is done properly and the training goes quite well.

Cheers

Pierre

**Comment by martinkersner** [comment]
Hi @pierrebaque,

I added suffix "-ft" in order to be able to fine tune on different dataset. That should be common practice. All those layers' weights and biases are filled within initialization. 

But maybe you are right. Thank you for the tip!

Martin

**Comment by bittnt** [comment]
@pierrebaque Her you go, the example CRF-RNN training scripts that works for VOC 2012 dataset. Make sure you have initialize the correct deconvolution layer correctly in the caffemodel file if you have different number of classes. https://github.com/torrvision/crfasrnn/tree/master/caffe-crfrnn/examples/semanticsegmentation_crfrnn

**Comment by pierrebaque** [comment]
@bittnt  Thank you, I'll have a look.

Cheers

**Comment by koizora223** [comment]
Hi @bittnt 
I use your trainning scripts to fine-tune TVG _CRFRNN_COCO.caffemodel on Pascal VOC datasets.
I trained 20 classes. But it doesn't work.There is an error on pasing NetParameter file. 
The error is : Error pasing text-format caffe.NetParameter:4:9:Expected inter or identifier.
What causes this error?
Thanks!

**Comment by bittnt** [comment]
@koizora223 Because you are not using the correct version of caffe. The error shows that the caffe cannot identify the type: "Data".

**Comment by koizora223** [comment]
@bittnt 
Thanks for your tranning scripts. I have solved my problem. I just modified the solver.prototext and and traintest.prototxt for fine-tunning. And it works now. 
Cheers.

**Comment by caffecoffee** [comment]
@martinkersner 
Hi! Thank you for the project! 
I trained a 3-classes model by finetuning the "TVG_CRFRNN_COCO_VOC.caffemodel" as README tells. But the output is awful!
My opinion is when output change 21 to 4,  some "Deconvolution"  layers and "Convolution" layers are abandoned. The parameter of these layers is  uninitialized.
To train a 3-classes model still trouble me. Do I should train a fcn-8s firstly? Then what? Does the finetune work? 
I am new about this. I am eager to get some instruction. I will appreciate any help!

**Comment by marticsvk** [comment]
@martinkersner @bittnt 
Yes exactly as you wrote @caffecoffee. It seems not working properly due to some issue with initialization when training with less classes than original 21. @pierrebaque you made successfully training only with full 21 classes, right?

**Comment by bittnt** [comment]
This could be fixed when you are initialising your decovolution layer properly. After you fix this in prototxt, It should work with any classes.
